Different protocols in the same Service definition with type=LoadBalancer

janosi opened the pull request:
This is the KEP for
kubernetes/kubernetes issue https://github.com/kubernetes/kubernetes/issues/23880
kubernetes/kubernetes PR https://github.com/kubernetes/kubernetes/pull/75831

kubernetes/enhancements issue: https://github.com/kubernetes/enhancements/issues/1435

janosi commented:
/assign @caseydavenport

sftim commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,351 @@
+
+Once that limit is removed those Service definitions will reach the Cloud Provider LB controller implementations. The logic of the particular Cloud Provider LB controller and of course the actual capabilities and architecture of the backing Cloud Load Balancer services determines how the actual exposure of the application really manifests. For this reason it is important to understand the capabil ...
+
+### User Stories [optional]
```
```sugestion
### User Stories
```

janosi replied:
Fixed.

sftim commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,351 @@
+
+As a Kubernetes cluster user I want to expose an application that provides its service on different protocols with a single cloud provider load balancer IP. In order to achieve this I want to define different `ports` with mixed protocols in my Service definition of `type: LoadBalancer`
+
+### Implementation Details/Notes/Constraints [optional]
```
```suggestion
### Implementation Details/Notes/Constraints
```

janosi replied:
Fixed.

sftim commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,351 @@
+
+### Option Control Alternatives
+
+#### Annotiation in the Service definition
```
```suggestion
#### Annotation in the Service definition
```

janosi replied:
Fixed

thinalai commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,351 @@
+The OpenStack CPI supports TCP, UDP, HTTP(S) in Service definitions and can configure the Octavia listeners with the protocols defined in the Service.
+OpenStack Octavia supports TCP, UDP and HTTP(S) on listeners, an own listener must be configured for each protocol, and different listeners can be used on the same LB instance.
+
+Summary: the OpenStack based clouds that use Octavia v2 as their LBaaS seems to support this feature once implemented. Pricing is up to their model.
```
[Octavia version](https://github.com/openstack/octavia/commit/b42a64a00ef549c69bc6835db9234a889b5ba62c) should >= 5.0.0 which first release in [OpenStack Train](https://releases.openstack.org/train/index.html#train-octavia)

janosi replied:
@thinalai Thank you for the comment! Indeed, I can see this bugfix in the [5.0.0 release notes ](https://docs.openstack.org/releasenotes/octavia/train.html#relnotes-5-0-0-stable-train)
`Fixed bug which prevented the creation of listeners for different protocols on the same port (i.e: tcp port 53, and udp port 53).`
I add this note to the KEP.

andrewsykim commented:
For reference, the initial issue/PRs that added validation against mixed protocols:
Issue: https://github.com/kubernetes/kubernetes/issues/20394
PR: https://github.com/kubernetes/kubernetes/pull/20600

andrewsykim commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,354 @@
+
+[GCP forwarding_rules_charges](https://cloud.google.com/compute/network-pricing#forwarding_rules_charges) suggest that the same K8s Service definition would result in the creation of 2 forwarding rules in GCP. This has the same fixed price up to 5 forwarding rule instances, and each additional rule results in extra cost.
+
+A user can ask for an internal TCP/UDP Load Balancer via a K8s Service definition that also has the annotation `cloud.google.com/load-balancer-type: "Internal"`. Forwarding rules are also part of the GCE Internal TCP/UDP Load Balancer architecture, but in case of Internal TCP/UDP Load Balancer it is not supported to define different forwarding rules with different protocols for the same IP addres ...
```
How would a user find out this feature isn't supported for this particular case? It sounds like either searching logs or realizing the LB is not serving the forwarding rules as expected?

janosi replied:
This is an interesting point. The current GCP CPI implementation takes the first protocol value it finds in the Service definition and uses that for all ports in the Service. There is no check for violations of the "single protocol" rule, i.e. it assumes that the filter in the K8s API already blocked a mixed protocol setup. 
If this filter is removed by this feature then - according to my understanding - the GCP CPI will setup an internal load balancer with the same protocol on all forwarding rules, even if the trigger was a Service with mixed protocol setup. There is no error feedback to the user.

thockin replied:
We do not have arbitrary status, but we could add it.  `status.loadBalancer` could grow a field or two.  Or we could use events.

thockin replied:
Open question

janosi replied:
Indeed. Thank you! 

Which solution would you prefer: the new field(s) in the `status.loadBalancer`or  the events? 

I do not see them as contradictory, though. Events are moved to external troubleshooting systems pretty fast, so maybe it is worth to keep the actual state of the Service in-cluster too.

janosi replied:
I added now this idea to the "Proposed solution" chapter.

bowei replied:
I would love to see improvements to the LB status fields if possible.
At the very least, there should be some form of conditions to allow for better extensibility.

janosi replied:
Do you mean users can use different protocols on internal GCE LBs now?
I assume, documentation update is also coming? Here it still states that multiple forwarding rules for the same IP can have either UDP or TCP, but not both: https://cloud.google.com/load-balancing/docs/internal#multiple_forwarding_rule
And backend services also has similar limitation: https://cloud.google.com/load-balancing/docs/internal#backend-service

Or it is only my bad English parser :)

thockin replied:
> Do you mean users can use different protocols on internal GCE LBs now?

Yes :)

I'll ping docs people.  See https://cloud.google.com/kubernetes-engine/docs/how-to/internal-load-balancing#enabling_shared_ip

janosi replied:
Fixed.

andrewsykim commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,354 @@
+A user can ask for an internal Load Balancer via a K8s Service definition that  has the annotation `service.kubernetes.io/qcloud-loadbalancer-internal-subnetid: subnet-xxxxxxxx`. Internal CLBs are free.
+
+
+### Risks and Mitigations
```
I think we also need to talk about upgrade/downgrade scenarios here. If I upgrade to a version that supports mixed protocol and then I downgrade to one that doesn't, will any of the existing cloud providers handle this unexpectedly? Sounds like worse case is dangling LBs because the old provider could assume single protocol listeners?

janosi replied:
@andrewsykim I added some thoughts about upgrade/downgrade. Please check!

andrewsykim commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,354 @@
+
+## Proposal
+
+The first thing proposed here is to lift the hardcoded limitation in Kubernetes that currently rejects Service definitions with different protocols if their type is LoadBalancer. Kubernetes would not reject Service definitions like this from that point:
```
`hardcoded limitation` -> `API validation`

janosi replied:
Fixed

janosi commented:
@andrewsykim 
> Can we squash the commits into more meaningful commits please?

Fixed. Squashed my internal (i.e. before PR creation) commits into one.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,392 @@
+- if a user is happy with 2 NLB (Service) instances for TCP and UDP still the user has two more forwarding rules to be billed - i.e. it has the same effect on pricing as if those TCP and UDP endpoints were behind the same NLB instance
+- already now the bills of users is affected if they have more than 5 forwarding rules as the result of their current Service definitions (e.g. 6 or more port definitions in a single Serice, or if the number of all port definitions in different Services is 6 or more, etc.)
+
+That is, if we consider the "single Service with 6 ports" case the user has to pay more for that single Service ("single LB instance") than for another Service (another LB instance) with 5 or less ports already now. It is not the number of LBs that matters. This phenomenon is already there with the current practice, and the enabling of mixed protocols will not change it to the worse.
```
This is not right?  We can (and do) set up a single Forwarding Rule for multiple ports.

janosi replied:
You are right, I was wrong here. Excuse me for this, fixed.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,392 @@
+
+The other risk is in the introduction of this feature without an option control mechanism, i.e. as a general change in Service handling. In that case there is the question whether this feature should be a part of the conformance test set, because it can affect the conformance of cloud providers.
+
+A possible mitigation is to put the feature behind option control. 
```
If we allow existing services to have this condition in 1.18 (e.g. implementing it and not rejecting updates to objects that already are in this state) but not allow new services (e.g. rejecting a create or an update that enters this state), then we can allow create in 1.19, knowing that rollback to 1.18 is safe.

I think?  Am I missing anything?

andrewsykim replied:
Seems right to me -- we followed the same approach when we added finalizers for Service LBs. The only thing to consider is the code path won't be vetted as much if we don't allow new services to use it at all, at the very least there should be an opt-in for mixed protocols so we can verify it works on v1.18 but we need to make sure users know the consequence of enabling (i.e. no rollback). Maybe we do need a feature gate?

thockin replied:
yeah, these changes are always nuanced.  I *think* the right thing is to allow existing uses and allow new uses IFF gate is enabled.  That puts anyone who uses it under the gate at rollback risk.

I have asked @liggitt for a consult, since he is involved in all of these :)

liggitt replied:
> I think the right thing is to allow existing uses and allow new uses IFF gate is enabled.

That's correct. The essential thing that default configurations (e.g. no opt into alpha feature gate) of version N does not allow new uses (via creation or update of an object that doesn't currently use the relaxed validation) that would be considered invalid by version N-1.

> That puts anyone who uses it under the gate at rollback risk.

That is why we put relaxed validation and new fields in GA objects under an alpha gate for one release, and limit API server rollback/skew to a single version.

thockin replied:
Strictly speaking - it doesn't HAVE to be "alpha", it could just be off-by-default beta (not much difference practically).

janosi replied:
Fixed. Let's see if it is better now. Please check.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,392 @@
+The implementation of the basic logic is ready in this PR:
+https://github.com/kubernetes/kubernetes/pull/75831
+
+Currently a feature gate is used to control its activation status. Though if we want to keep this feature behind option control even after it reaches its GA state we should come up with a different solution, as feature gates are used to control the status of a feature as that graduates from alpha to GA, and they are not meant for option control for features with GA status.
```
I'm not sure an alpha is warranted here, but maybe.  It's not a huge change, except in the cloud providers it could be.  Thoughts?

andrewsykim replied:
Yeah I think we skip alpha as long as rollbacks don't break LBs as you mentioned here https://github.com/kubernetes/enhancements/pull/1438#discussion_r367675751

janosi replied:
Do I understand right then that the agreed change is the following:
- in 1.18 creating a mixed Service is allowed if the feature flag is ON. We add notification to the documentation that in this case rollback is not possible, or at least can break things depending on the cloud provider implementation
- in 1.18 we allow update of such Services even if the feature flag in OFF
- in 1.19 we change the create path so, that even if the feature flag is OFF mixed Services can be created
?
Thank you!

liggitt replied:
> in 1.18 creating a mixed Service is allowed if the feature flag is ON. We add notification to the documentation that in this case rollback is not possible, or at least can break things depending on the cloud provider implementation

That's correct, and that means the feature gate should be alpha and disabled by default (we should not enable something by default that breaks skew/rollback).

> in 1.18 we allow update of such Services even if the feature flag in OFF

Correct

> in 1.19 we change the create path so, that even if the feature flag is OFF mixed Services can be created
?

Typically, we would leave the create/update logic the same as in 1.18, and simply promote the feature flag to beta and enable it by default in 1.19. That gives a release to get broader feedback and bug reports, but allows someone to turn off the feature and stop allowing new uses if problems are found, without cutting a new Kubernetes 1.19.x patch release.

Once the feature graduates to GA and is locked on (e.g. in 1.20), *then* all feature-gate-conditional logic can be dropped.

janosi replied:
@liggitt @thockin @andrewsykim I think I updated the document according to this logic. Please check and suggest how to continue.

thockin commented:
That makes sense.

On Thu, Jan 16, 2020, 3:24 PM Andrew Sy Kim <notifications@github.com>
wrote:

> *@andrewsykim* commented on this pull request.
> ------------------------------
>
> In keps/sig-network/20200103-mixed-protocol-lb.md
> <https://github.com/kubernetes/enhancements/pull/1438#discussion_r367701090>
> :
>
> > +Our feature does not introduce new values or new fields. It enables the usage of an existing value in existing fields, but with a different logic. I.e. if someone creates a Service with mixed protocol setup and then rollbacks the API server to a version that does not implement this feature the clients will still get the Service with mixed protocols when they read that via the rollback'd API. If the client (CPI implementation) has been rollbacked, too, then the client may receive such a Service setup that it does not support.
> +
> +- Alibaba: no risk. The current CPI and LB already supports the mixed protocols in the same Service definition. If this feature is enabled in an API server and then the API server rollback is executed the CPI can still handle the Services with mixed protocol sets.
> +- AWS: no risk. The current CPI accepts Services with TCP protocol only, i.e. after a K8s upgrade a user still cannot use this feature. Consequently, a rollback in the K8s version does not introduce any issues.
> +- Azure: no risk. The current CPI and LB already supports the mixed protocols in the same Service definition. The situation is the same as with the Alibaba CPI.
> +- GCE: currently the GCE CPI assumes that a Service definition contains a single protocol value, as it assumes that the Service Controller already rejected Services with mixed protocols. While the Service Controller really did so a while ago, it does not do this anymore. It means a risk.
> +- IBM Cloud VPC: no risk. The same situation like in the case of AWS.
> +- IBM Cloud Classic: no risk. The CPI and NLB already supports TCP and UDP in the same Service definition. The same situation like in the case of Alibaba.
> +- OpenStack: no risk. The CPI and NLB already supports TCP and UDP in the same Service definition. The same situation like in the case of Alibaba.
> +- Oracle: no risk. The CPI and LB already supports mixed protocols. The same situation like in the case of Alibaba.
> +- Tencent Cloud: no risk. The CPI and LB already supports mixed protocols. The same situation like in the case of Alibaba.
> +
> +
> +The other risk is in the introduction of this feature without an option control mechanism, i.e. as a general change in Service handling. In that case there is the question whether this feature should be a part of the conformance test set, because it can affect the conformance of cloud providers.
> +
> +A possible mitigation is to put the feature behind option control.
>
> Seems right to me -- we followed the same approach when we added
> finalizers for Service LBs. The only thing to consider is the code path
> won't be vetted as much if we don't allow new services to use it at all, at
> the very least there should be an opt-in for mixed protocols so we can
> verify it works on v1.18 but we need to make sure users know the
> consequence of enabling (i.e. no rollback). Maybe we do need a feature gate?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/enhancements/pull/1438?email_source=notifications&email_token=ABKWAVF37SUFCU2P4BL4LV3Q6DUEBA5CNFSM4KEOJSSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCSCVH5Q#discussion_r367701090>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVCZMVNVBP6FVG6PTXLQ6DUEBANCNFSM4KEOJSSA>
> .
>

janosi commented:
@caseydavenport @thockin @andrewsykim Do you think there is a chance to get it into 1.18 enhancements? Or do you see some gaps here? Thank you!

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,382 @@
+ - a feature flag shall control whether loadbalancer Services with mixed protcol configuration are accepted or not at validation time
+ - we must add a note to the documentation that if such Service is created then it may break things after a rollback - it depends on the cloud provider implementation
+ - the update of such services shall be possible even if the feature flag is OFF. This is to prepare for the next release when the feature flag is removed from the create path, too, and after a rollback to the first release the update of existing Service objects must be possible
+ - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. It does not mean that the CPI implementations have to accept such Services. It means that the CPI implementations shall be able to process such Services. The CPI implementation can still reject such a Service if it cannot manage the corresponding cloud load balancers for those Services, or for a ...
```
What does it mean to "reject" a service?

janosi replied:
Well, just like currently some CPIs do not accept a service if for example not all ports are TCP.  My assumption was that that is a kind of accepted way how a CPI can "reject" a service. Similarly, the CPI can still "reject" the service on the same way if it does not want to support this feature.

janosi replied:
Reworded.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,382 @@
+#### Proposed solution
+
+In the first release: 
+ - a feature flag shall control whether loadbalancer Services with mixed protcol configuration are accepted or not at validation time
```
The gate covers *new* uses.

janosi replied:
Fxied.

janosi commented:
@caseydavenport @thockin @andrewsykim Would you like to see some more updates on this KEP?

janosi commented:
@thockin event though we missed the 1.18 deadline with this one, I would like to progress. Please advise what is next. If the KEP is good to go, I can update the necessary meta fields, and it can be merged then. If there are still things to update in it, I would like to know those.
Also I wonder if this KEP could be a candidate for an "exception request" so we could still have it in 1.18.

trinitronx commented:
Hope this helps with triage:

In 2019, this feature request was apparently added to milestones: [v1.17][1], [v1.18][2] in kubernetes/kubernetes#80394

Other PRs and Issues seem related:

- Related PRs: kubernetes/kubernetes#20600, kubernetes/kubernetes#75831, kubernetes/kubernetes#80394,  kubernetes/kubernetes#87549
- Related issues: kubernetes/kubernetes#23880, kubernetes/enhancements#1435, kubernetes/enhancements#1438

[1]: https://github.com/kubernetes/kubernetes/pull/80394#issuecomment-528340593
[2]: https://github.com/kubernetes/kubernetes/pull/80394#issuecomment-553227338

andrewsykim commented:
/ok-to-test

danwinship commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,382 @@
+
+#### Story 1
+
+As a Kubernetes cluster user I want to expose an application that provides its service on different protocols with a single cloud provider load balancer IP. In order to achieve this I want to define different `ports` with mixed protocols in my Service definition of `type: LoadBalancer`
```
This isn't a user story; it's just the KEP Summary with "As a user I want" prepended to it. :stuck_out_tongue: 

A user story should give a *concrete* example of something users want to do that they can't do currently, to give readers a solid understanding of why this KEP is (or maybe isn't) a good idea.

Eg, (totally making up a fake story based on the example in the Proposal section above), "I want to use Kubernetes to provide DNS service by exposing a LoadBalancer that does DNS over both TCP and UDP so that other DNS servers and hosts can make DNS requests to it regardless of whether they use TCP or UDP".

(I imagine you probably have some SCTP-based example of something users want to do with this?)

janosi replied:
Yeah. 
Even though I have some part in the SCTP support feature, and really nothing else in K8s, indeed, this time I have not started dealing with this KEP due to SCTP.
Updated the user stories (well, let's hope these can be called user stories now)

danwinship commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,382 @@
+
+#### Billing perspective
+
+The goal of the current restriction on the K8s API was to prevent an unexpected extra charging for Load Balancers that were created based on Services with mixed protocol definitions.
```
Maybe mention that earlier; I was kind of confused when you suddenly starting talking about billing. I had assumed that the problem was solely going to be that some clouds supported multiprotocol and others didn't (eg like AWS)

janosi replied:
Yes. Because as I learnt, the main reason to disable this feature explicitly was to prevent surprising/shocking doubled LB bills in clouds where the billing happens per protocol. That is why the main direction was to show how such LBs would be charged at the different cloud providers if this feature were re-enabled.

janosi replied:
Fixed.

andrewsykim commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,382 @@
+- IBM Cloud Classic: no risk. The CPI and NLB already supports TCP and UDP in the same Service definition. The same situation like in the case of Alibaba.
+- OpenStack: no risk. The CPI and NLB already supports TCP and UDP in the same Service definition. The same situation like in the case of Alibaba.
+- Oracle: no risk. The CPI and LB already supports mixed protocols. The same situation like in the case of Alibaba.
+- Tencent Cloud: no risk. The CPI and LB already supports mixed protocols. The same situation like in the case of Alibaba.
```
@timoreimann can you comment for the DigitalOcean provder?

timoreimann replied:
The DigitalOcean CPI (Cloud Provider Interface) only supports TCP on the Service definition (i.e., no UDP). Using Service annotations, the cloud load balancer can be configured to use TCP, HTTP, HTTPS, and HTTP2.

I think that makes DigitalOcean fall under the "no risk" category since we do not support more than one L4 protocols (TCP).

janosi replied:
Fixed.

janosi commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,382 @@
+  - turns off the feature flag; or
+  - executes K8s version rollback to the N-1 version where this feature is not available at all
+
+When investigating the possible issues with such a change we must consider [the supported version skew among components](https://deploy-preview-11060--kubernetes-io-master-staging.netlify.com/docs/setup/version-skew-policy/), which are the K8s API server and the cloud controller managers (CPI implementations) in our case.
```
Fixed.

janosi commented:
@thockin @danwinship @andrewsykim I would appreciate your opinion/comments.

sftim commented:
/uncc
Previous feedback was just nits

danwinship commented:
I don't really know a lot about cloud load balancer pricing, so I don't have much more to say. (Though my earlier comment about "you should mention that this is about money sooner" still stands; none of "Summary", "Motivation", or "Goals" mentions billing concerns.)

janosi commented:
@danwinship 
> Though my earlier comment about "you should mention that this is about money sooner" still stands; none of "Summary", "Motivation", or "Goals" mentions billing concerns.

You are right. I added some explanation to the "Motivation" part now.

I hope we can progress with this one.

danwinship commented:
> My concern with only having a feature gate is that it's still difficult for users to discover what protocols their cloud provider actually supports.

How much is that *kubernetes*'s problem though? I would hope that users would investigate what features their cloud provides *before* they start paying them...

> In the future it would be nice if cloud providers can support validating webhooks for this kind of thing but we're too busy at the moment just trying to get them out-of-tree.

The validation could be in the controller-manager, powered by some sort of `GetLoadBalancerSupportFlags` call on the cloud provider. (eg, the controller manager would call that at startup, and the cloud provider would say "I support TCP, UDP, and SCTP, in any combination; and IPv4 and dual-stack, but not IPv6-single-stack", and then the controller manager could validate services based on that.)

andrewsykim commented:
> The validation could be in the controller-manager, powered by some sort of GetLoadBalancerSupportFlags call on the cloud provider. (eg, the controller manager would call that at startup, and the cloud provider would say "I support TCP, UDP, and SCTP, in any combination; and IPv4 and dual-stack, but not IPv6-single-stack", and then the controller manager could validate services based on that.)

It should also be out-of-tree so it would be on the cloud-controller-manager specifically. This seems like the most ideal solution but admittedly it's a lot more work.  

> How much is that kubernetes's problem though? I would hope that users would investigate what features their cloud provides before they start paying them...

Ideally yes but in reality the two are still coupled too much.

janosi commented:
> It should also be out-of-tree so it would be on the cloud-controller-manager specifically. This seems like the most ideal solution but admittedly it's a lot more work.

I checked the CCMs 2 months ago, but there are definitely checks for the list of protocols and the CCM stops creating the LB if the protocol list in the Service request contains unsupported protocol value.

> My concern with only having a feature gate is that it's still difficult for users to discover what protocols their cloud provider actually supports.

I agree with @danwinship Users should check the supported protocols. I mean, if UDP is not supported for example then the user who creates a Service with protocol=UDP does not get the LB. That is already the current functionality. I wonder if this feature would make the situation worse from UX perspective.

andrewsykim commented:
> I checked the CCMs 2 months ago, but there are definitely checks for the list of protocols and the CCM stops creating the LB if the protocol list in the Service request contains unsupported protocol value. 

I don't think these checks are adding much value because Services (type=LB) with UDP ports would fail API validation right? 

> I mean, if UDP is not supported for example then the user who creates a Service with protocol=UDP does not get the LB. That is already the current functionality. I wonder if this feature would make the situation worse from UX perspective.

A user getting an error at API validation because UDP is not supported vs digging through logs/events/console to figure it out is a big difference in the UX.

janosi commented:
> I don't think these checks are adding much value because Services (type=LB) with UDP ports would fail API validation right?

No. These checks are not there in the CCMs to prevent mixed protocol in a Service definition. Those are there to block such LB request that wants to use an unsupported protocol.
Example: if someone creates now a Service with Protocol=UDP only and passes it to a cluster running on AWS the AWS CCM will not allow the creation of the LB because there is a check in the CCM that only Services with TCP protocol are supported. I.e. they do not block a Service with TCP+UDP. They block any Service with any other protocol than TCP.

> A user getting an error at API validation because UDP is not supported vs digging through logs/events/console to figure it out is a big difference in the UX IMHO.

I am aware of that and I agree.  Though then we start talking about much more here than supporting mixed protocols in Services. Then we start talking about how a cloud specific check could be added to the API validation. Which was not the purpose here. People in the related issue and in the PR are waiting for this change to finally be able to use their services that would require 2 different protocols on the same LB.

andrewsykim commented:
> People in the related issue and in the PR are waiting for this change to finally be able to use their services that would require 2 different protocols on the same LB.

In https://github.com/kubernetes/enhancements/pull/1438#pullrequestreview-383948229 I mentioned how I would prefer to move forward with this today without the additional complexity discussed around validation webhooks: an apiserver flag specifying Service type=LB protocols supported (something like `--service-loadbalancer-protocols`) which determines validation behavior. In doing so we reduce the # of people that need to care about this configuration (ideally only the cluster administrators) and any other consumer of the API will be gated by the API validation for using unsupported protocols. 

Just my own 2 cents on this, approval ultimately up SIG Network leads :)

janosi commented:
I tried to think about that API server flag. I wonder how it would work. Let's say currently a cloud supports both UDP and TCP as LB protocols (of course not via the same Service definition, but in general).  Once we introduce that flag, should the cloud provider immediately set this API server flag to "tcp,udp"? Or do we define a set of protocols that are allowed by default? E.g. TCP+UDP? But then what about those clouds that so far did not support UDP? Then this default TCP+UDP would mean we allow wrong set of protocols in those clouds, which is bad for the UX.
Then if this flag enables the usage of different protocols in a Service definition starting from release X, what happens when the user creates such a Service definition in release X, and then falls back to release X-1 for any reasons - this is the use case due to which the feature flag based solution became so complex. This API server flag is not present in release X-1, so in the end we will need the same release mechanism that we wanted to introduce with the feature flag.
Maybe I just do not understand clearly this API server flag. But as it starts controlling something on the API, doesn't it have the same backward compatibility issue as we have with the feature flag?

andrewsykim commented:
>  Once we introduce that flag, should the cloud provider immediately set this API server flag to "tcp,udp"? Or do we define a set of protocols that are allowed by default? E.g. TCP+UDP? But then what about those clouds that so far did not support UDP? Then this default TCP+UDP would mean we allow wrong set of protocols in those clouds, which is bad for the UX.

It would default to TCP only. Opting into UDP would require action from the user to update that flag. The key difference is that a user setting `--service-loadbalancer-protocols=tcp,udp` (likely) means they've checked that their platform supports UDP ports (we can add pre-warning in the flag description to ensure they do). Setting a catch all feature flag like `ServiceLoadBalancerMixedProtocol` is a bit ambiguous and a user could just set it and assume they can use UDP ports. Fwiw maybe I'm overthinking this, would love to hear other people's opinions (cc @cheftako).

> Then if this flag enables the usage of different protocols in a Service definition starting from release X, what happens when the user creates such a Service definition in release X, and then falls back to release X-1 for any reasons - this is the use case due to which the feature flag based solution became so complex. This API server flag is not present in release X-1, so in the end we will need the same release mechanism that we wanted to introduce with the feature flag.

Similar to what is already proposed, we need cloud providers to support existing Services w/ UDP ports 1 release before enabling this feature so we can handle Services w/ UDP ports from a version we rolled back from. In this case we could add the `--service-loadbalancer-protocols` 1 - 2 versions after we know most (if not all) providers support UDP ports.

janosi commented:
@cheftako Could you please comment? As I understand @andrewsykim would like your opinion on this one. Thank you!

edobry commented:
Hello, how much work is left on this PR? Would love to get closer to having this functionality. Happy to contribute if there is a way to do so!

janosi commented:
@edobry It depends on which alternative in the KEP will be accepted. I take the KEP to the SIG-Network meeting next week, and I hope it can get an approval soon.

bowei commented:
/assign

M00nF1sh commented:
this is valuable for AWS as well, as it can use TCP/UDP to support use case like DNS server

adampl commented:
I don't envision a bright future for Kubernetes if such trivial issues take over half a year to develop a _proposal_, 4 years since the heavily-supported feature request has been posted. I'm afraid this project is slowly drowning in corporate bureaucracy...

andrewsykim commented:
@adampl please do not trivialize the discussion here. Some of these changes are more nuanced than they appear and we do our best not to break users. 

As I've mentioned previously in this thread, I think a validating admission webhook from the cloud provider (in cloud-controller-manager) is the right solution given the tight coupling of Service Type=LoadBalancer and the cloud provider. I would be in favor of just removing the current mixed protocol validation gate after the webhook in place. This would also remove the need for a feature gate or an apiserver flag.

Whether we should remove the existing validation for mixed protocol before we have the validating webhook in place is not up to me but I'm not in favor of it. Will defer to SIG Network leads on that one.

adampl commented:
@andrewsykim Sorry for the interruption, but it's really frustrating to watch. Kubernetes should be pushing the progress forward, instead of waiting for the big sluggish cloud providers to catch up. They are free not to support the latest version if they're incapable of implementing its requirements. The current approach is not working for the benefit of the broader community of k8s users.

TBBle commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,399 @@
+Our feature does not introduce new values or new fields. It enables the usage of an existing value in existing fields, but with a different logic. I.e. if someone creates a Service with mixed protocol setup and then rollbacks the API server to a version that does not implement this feature the clients will still get the Service with mixed protocols when they read that via the rollback'd API. If t ...
+
+- Alibaba: no risk. The current CPI and LB already supports the mixed protocols in the same Service definition. If this feature is enabled in an API server and then the API server rollback is executed the CPI can still handle the Services with mixed protocol sets.
+- AWS: no risk. The current CPI accepts Services with TCP protocol only, i.e. after a K8s upgrade a user still cannot use this feature. Consequently, a rollback in the K8s version does not introduce any issues.
```
https://github.com/kubernetes/kubernetes/pull/92109 will introduce UDP support to the AWS CPI. It's been approved, and will hopefully be merged soon. The updated CPI will _not_ work correctly with mixed TCP/UDP without further work, c.f. https://github.com/kubernetes/kubernetes/pull/92109#discussion_r439730341

janosi replied:
@TBBle Fixed the AWS related parts.

TBBle commented:
Was there any movement on this? I watched the relevnt part of the SIG-Network meeting recording from 11th June, and someone there agreed to review this, but I did not see any followup in later SIG-Network meetings or their mailing list.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,399 @@
+The implementation of the basic logic is ready in this PR:
+https://github.com/kubernetes/kubernetes/pull/75831
+
+Currently a feature gate is used to control its activation status. Though if we want to keep this feature behind option control even after it reaches its GA state we should come up with a different solution, as feature gates are used to control the status of a feature as that graduates from alpha to GA, and they are not meant for option control for features with GA status.
```
I don't think we want to make this an optional feature long-term, just as it is introduced

janosi replied:
Fixed. Hopefully better articulated this time.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,399 @@
+ - a feature flag shall control whether new loadbalancer Services with mixed protcol configuration can be created or not
+ - we must add a note to the documentation that if such Service is created then it may break things after a rollback - it depends on the cloud provider implementation
+ - the update of such Services shall be possible even if the feature flag is OFF. This is to prepare for the next release when the feature flag is removed from the create path, too, and after a rollback to the first release the update of existing Service objects must be possible
+ - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. It does not mean that the CPI implementations have to accept such Services. It means that the CPI implementations shall be able to process such Services. The CPI implementation can still validate the Service objects to check whether any of those has mixed protocol configuration and if the CPI ca ...
```
Let's tighten this spec.

1) Once the apiserver allows mixed-protocol Service specs, each CPI has to decide if and how to support it.
2) If a CPI chooses to support mixed-protocol, everything is good
3) If a CPI can't support mixed-protocol, should they a) create the ones the support; b) fail the whole thing?
4) If a CPI can't support mixed-protocol, they must log that in new status fields (need a proposal for schema) and they should deliver an event

thockin replied:
Unfortunately, if a CPI does not support mixed-protocol LB and doesn't set status (e.g. has not been updated), there's not much we can do.

janosi replied:
Fixed. Tried to be more straight about the requirements. Also added the schema proposal.

andrewsykim replied:
I'm exploring how much work is involved to have the CPI setup a validating webhook for Services Type=LB, so we can have provider-specific validation of port protocols. 

Hopefully it's available around the same time we lift this validation but don't think we need to block on it.

thockin commented:
Obnoxious, but:  The KEP format has changed to a directory.  I'm fine to merge this and then convert it.

TBBle commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -340,17 +341,79 @@ Con:
  - we must add a note to the documentation that if such Service is created then it may break things after a rollback - it depends on the cloud provider implementation
  - the update of such Services shall be possible even if the feature flag is OFF. This is to prepare for the next release when the feature flag is removed from the create path, too, and after a rollback to the first release the update of existing Service objects must be possible
- - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. It does not mean that the CPI implementations have to accept such Services. It means that the CPI implementations shall be able to process such Services. The CPI implementation can still validate the Service objects to check whether any of those has mixed protocol configuration and if the CPI ca ...
+ - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. Either via supporting such Service definitions, or clearly indicating to the users that the Service could not be processed as specified. As we can see from our analysis some CPIs support other protocols than TCP and UPD in the Service definitions, while others support only TCP and UDP. That is t ...
```
Typo: "UPD" instead of "UDP"

janosi replied:
Fixed.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -340,17 +341,80 @@ Con:
+      "description": "The protocol for this port",
+      "type": "string"
+    },
+    "name": {
```
I don't think we need name in status

janosi replied:
Fixed.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -340,17 +341,80 @@ Con:
+      "description": "The name of this port within the service.",
+      "type": "string"
+    },
+    "ready": {
```
Are there more states that a single port or set of ports might be in? I'm wondering if bool is right or enum or even Conditions.  Bool is simple, maybe good enough...

janosi replied:
I have been thinking about it, but could not find more states for the time being.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -340,17 +341,80 @@ Con:
+      "description": "Specifies whether the port was configured for the load-balancer.",
+      "type": "boolean"
+    },
+    "message": {
```
We don't have a lot of precedent on this sort of thing, other than Conditions.  Perhaps simply mapping ports here, and adding Conditions to status would be more expressive.  E.g.:

```
conditions:
 - type: LoadBalancerMixedProtocolNotSupported
```

janosi replied:
I created a "conditions" field right into "Service.status.loadBalancer", and defined the first condition with this name.

TBBle replied:
While we're at it, should there be a condition for just _this_ protocol is not supported? i.e. for CPI that can do LB mixed protocols, but the protocol for this port cannot work, either because it's flat-out not supported, e.g., UDP on AWS ELB, or because for some reason the Load Balancer cannot be built with multiple protocols when this one is present.

At least in the AWS UDP-on-ELB case, this would be better than just an event saying 'We couldn't do that' as we get now. The event still carries _details_, but being able to set a condition would be nice too.

janosi replied:
Do you think the new "portStatuses" could be used to indicate problems with the particular ports? I.e. the "Service.status.conditions" would show problem with the mixed protocol setup, and then the "ingress.portStatuses" could be used to provide more information.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -340,17 +341,80 @@ Con:
+      },
+      "type": "array"
+    },
+    "portStatuses": {
```
I think we want `ports` under `ingress` - ingress is a plural on purpose, so ports should be tracked based on each ingress point.

janosi replied:
Fixed, moved the portStatuses under ingress.

TBBle commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,463 @@
+  "description": "LoadBalancerStatus represents the status of a load-balancer.",
+  "properties": {
+    "ingress": {
+      "description": "Ingress is a list containing ingress points for the load-balancer. Traffic intended for the service ld be sent to these ingress points.",
```
"ld"?

janosi replied:
Hmmm, bad copy-paste. Fixed.

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -405,6 +428,28 @@ In the first release:
+      },
+      "type": "array"
+    },
+    "conditions": {
```
Generally conditions are just under Status since they represent potentially other facets of the resource.  I don't know if we have any precedent for nesting them further - do you think this is really a valuable trend to set?  It would be simpler and more consistent to define them higher at `status`.

janosi replied:
OK, updated. Let's see if I got it right now :)

thockin commented on keps/sig-network/20200103-mixed-protocol-lb.md:
````diff
@@ -351,9 +351,35 @@ In the first release:
+ In order to provide a way for the CPI to indicate port statuses to the user we would add the following new `portStatus` list of port statuses to `Service.status.loadBalancer.ingress`, so the CPI can indicate the status of the LB. Also we add a `conditions` list of conditions to the `Service.status.loadBalancer`, with the first official condition `LoadBalancerMixedProtocolNotSupported` defined.
 
 ```json
+"io.k8s.api.core.v1.LoadBalancerCondition": {
````
I have to say, it's considerably harder to review this as openAPI than as Go code :)

janosi replied:
Yep, it sounded like a good choice for API specification. 😄 +1 that Go would be easier 😄

andrewsykim commented on keps/sig-network/20200103-mixed-protocol-lb.md:
```diff
@@ -0,0 +1,518 @@
+  ],
+  "type": "object"
+},
+"io.k8s.api.core.v1.PortStatus": {
```
Is kube-proxy expected to consume this in any way into iptables/IPVS rules or is this just for reporting? Might be worth clarifying this

thockin replied:
Good question.  If IPVS and iptables set up captures for the LBVIP + proto + port, and the LB itself doesn't support mixed protocols, should kube-proxy also not capture them?

janosi replied:
I wonder about the exact handling. I mean, no rules at all, or rather drop those packets that are anyway not handled by the LB on the external interface?

janosi replied:
So, I added a kube-proxy chapter. Please check and if you are OK with that please remove the hold flag from this one.

thockin commented:
I'm going to Approve this now, but I like Andrew's question.  We can either answer it as a followup or as one more rev on this.  Remove the hold as you see fit.

Also, I set the "squash" merge option for this, so you don't need to manually squash.

/approve
/lgtm
/hold

janosi commented:
...ahh, and a new lgtm if you are still OK with this one.

thockin commented:
/lgtm

hold removed
